File: cpp/16. 3Sum Closest.cpp
/*
2016-8-30 00:01:06

https://leetcode.com/problems/3sum-closest/

Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

    For example, given array S = {-1 2 1 -4}, and target = 1.

    The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).

*/

#include <iostream>
#include <algrithm>
#include <limits>
#include <math>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        
		sort(nums.begin(), nums.end());
		int minGap = numeric_limits<int>::max();
		int res = 0;
		
		for (auto i = nums.begin(); i < nums.end() - 2; ++ i)
		{
			auto beg = i + 1;
			auto end = nums.end() - 1;
			
			while (beg < end)
			{
				int sum = *i + *beg + *end;
				int gap = abs(target - sum);
				
				if (gap < minGap)
				{
					minGap = gap;
					res = sum;
				}
				
				if (sum < target)
				{
					++ beg;
				}
				else
				{
					-- end;
				}
			}
			
		}
		
		return res;
    }
};

File: java/src/main/java/club/cser/leetcode/16. 3Sum Closest.java
package club.cser.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class ThreeSumClosest {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        //  You may assume that each input would have exactly one solution.
        int resGap =  nums[0] + nums[1] + nums[2] - target;

        for (int i = 0; i < nums.length - 2; ++ i) {
            int left = i + 1;
            int right = nums.length - 1;

            // 重复数字 跳过
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            while (left < right) {

                int tmpGap = nums[i] + nums[left] + nums[right] - target;

                if (Math.abs(resGap) > Math.abs(tmpGap))
                    resGap = tmpGap;

                if (tmpGap < 0) {
                    ++ left;
                } else if (tmpGap > 0) {
                    -- right;
                } else {
                    return target;
                }
            }
        }

        return resGap + target;
    }
}